// @docusaurus/plugin-content-docs/lib/globalData.js
"use strict";
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.toGlobalDataVersion = exports.toGlobalSidebars = exports.toGlobalDataGeneratedIndex = exports.toGlobalDataDoc = void 0;
const lodash_1 = require("lodash");
const utils_1 = require("@docusaurus/utils");
const utils_2 = require("./sidebars/utils");
function toGlobalDataDoc(doc) {
    return {
        id: doc.unversionedId,
        path: doc.permalink,
        sidebar: doc.sidebar,
    };
}
exports.toGlobalDataDoc = toGlobalDataDoc;
function toGlobalDataGeneratedIndex(doc) {
    return {
        id: doc.slug,
        path: doc.permalink,
        sidebar: doc.sidebar,
    };
}
exports.toGlobalDataGeneratedIndex = toGlobalDataGeneratedIndex;
function toGlobalSidebars(sidebars, version) {
    const { getFirstLink } = (0, utils_2.createSidebarsUtils)(sidebars);
    return (0, lodash_1.mapValues)(sidebars, (sidebar, sidebarId) => {
        const firstLink = getFirstLink(sidebarId);
        if (!firstLink) {
            return {};
        }
        return {
            link: {
                path: firstLink.type === 'generated-index'
                    ? (0, utils_1.normalizeUrl)([version.versionPath, firstLink.slug])
                    : version.docs.find((doc) => doc.id === firstLink.id || doc.unversionedId === firstLink.id).permalink,
                label: firstLink.label,
            },
        };
    });
}
exports.toGlobalSidebars = toGlobalSidebars;
function toGlobalDataVersion(version) {
    return {
        name: version.versionName,
        label: version.versionLabel,
        isLast: version.isLast,
        path: version.versionPath,
        mainDocId: version.mainDocId,
        docs: version.docs
            .map(toGlobalDataDoc)
            .concat(version.categoryGeneratedIndices.map(toGlobalDataGeneratedIndex)),
        sidebars: toGlobalSidebars(version.sidebars, version),
    };
}
exports.toGlobalDataVersion = toGlobalDataVersion;
